lib/rack/cache/meta_store.rb

```diff
@@ -64,6 +64,18 @@ def store(request, response, entity_store)
         end
         response.headers['X-Content-Digest'] = digest
         response.headers['Content-Length'] = size.to_s unless response.headers['Transfer-Encoding']
+
+        # If the entitystore backend is a Noop, do not try to read the body from the backend, it always returns an empty array
+        unless entity_store.is_a? Rack::Cache::EntityStore::Noop
+          # A stream body can only be read once and is currently closed by #write.
+          # (To avoid having to keep giant objects in memory when writing to disk cache
+          # the body is never converted to a single string)
+          # We cannot always reply on body to be re-readable,
+          # so we have to read it from the cache.
+          # BUG: if the cache was unable to store a stream, the stream will be closed
+          #      and rack will try to read it again, resulting in hard to track down exception
+          response.body = entity_store.open(digest) || response.body
+        end
```

Comment:
What about replacing the whole conditional with this?

```
stored_body = entity_store.open(digest)
response.body = stored_body unless stored_body.nil? || stored_body.size.zero? 
```

This would remove checking for the Noop entitystore, which is ugly and imho should not be here.

```unless stored_body.nil? || stored_body.size.zero?```  is the simplest one-liner I can think of that works both with Array and File objects (```Rack::Cache::EntityStore::Disk::Body``` extends ```File```). It would preserve the original response body if it cannot be read from the cache, and Noop would keep working.